community/kernel/src/test/java/org/neo4j/test/TestGraphDatabaseFactory.java

```diff
@@ -95,17 +97,47 @@ protected TestGraphDatabaseFactoryState getStateCopy()
         return new TestGraphDatabaseFactoryState( getCurrentState() );
     }
 
-    public FileSystemAbstraction getFileSystem()
+    public Provider<FileSystemAbstraction> getFileSystem()
     {
         return getCurrentState().getFileSystem();
     }
 
-    public TestGraphDatabaseFactory setFileSystem( FileSystemAbstraction fileSystem )
+    /**
+     * Assigns a custom file system, which created graph databases will use.
+     *
+     * IMPORTANT to note is that the caller of this method is responsible for
+     * {@link FileSystemAbstraction#close() closing} this file system when this graph database factory is done
+     * and should be used no more. Reason is that a {@link GraphDatabaseFactory} can create multiple database
+     * instances and sometimes it's the intention that each database instance has the same file system,
+     * without having it closed.
+     */
+    public TestGraphDatabaseFactory setFileSystem( final FileSystemAbstraction fileSystem )
+    {
+        return setFileSystem( new Provider<FileSystemAbstraction>()
+        {
+            @Override
+            public FileSystemAbstraction instance()
+            {
+                // Knowing that the graph database will close this file system in GDS#shutdown()
+                // and also knowing that we'd like to keep this file system up and running until
+                // closed externally, potentially after multiple database instances have lived with
+                // this file system.
+                return new NonClosingFileSystemAbstraction( fileSystem );
+            }
+        } );
+    }
+
+    /**
+     * Assigns a custom file system {@link Provider}, which created graph databases will use, or rather
+     * whatever this provider {@link Provider#instance() provides} every time a new database is instantiated.
+     */
+    public TestGraphDatabaseFactory setFileSystem( Provider<FileSystemAbstraction> fileSystem )
```

Comment:
Opted to not use it because we may want to backport this to 2.2, or even further (although not with JimFS as _default_).